Add tabular regression example

lazarust opened the pull request.

adrinjalali commented:
I'm okay with adding this, but it'd need to add a few things for the example to be more useful:

- use skops format `skops.io` instead of pickle
- Have a story behind the example. It should be something where we could also explain limitations of the model, and disparities if they exist, to highlight how the model card should be used and written.
- Add a more detailed model card, with all sections filled and add a few different inspection and interpretation methods to better explain the model

Thanks for the work @lazarust

lazarust commented:
Hi @adrinjalali, thanks for taking a look at my PR! I've addressed your first point and replaced using pickle with `skops.io`. I did have some questions about your other two issues though. I tweaked the model text and comments but was unsure what sections I was missing. I based this example off of `examples/plot_text_classification.py`. If you could let me know what else you want to be added I can go ahead and add it and update the text classification example too. Thanks!

adrinjalali commented on examples/plot_tabular_regression.py:
```diff
@@ -0,0 +1,148 @@
+# Inference
+# =========
+# Let's see if the model works.
+prediction_data = [[100, 200, 300, 400, 500, 600, 700, 800, 900, 1000]]
```
this should use a few rows from `X_test` instead.

adrinjalali commented on examples/plot_tabular_regression.py:
```diff
@@ -0,0 +1,148 @@
+# to push your models please refer to
+# :ref:`this example <sphx_glr_auto_examples_plot_hf_hub.py>`.
+
+model_card.save(Path(local_repo) / "README.md")
```
missing new line at the end of the file.

adrinjalali reviewed (commented):
Thanks,

This should also add the generating file the same way as it's done here:

https://github.com/skops-dev/skops/blob/495abe678b7067d490d2703b6b3ec5176427b0cf/examples/plot_hf_hub.py#L85-L87

lazarust commented:
@adrinjalali I've made those requested changes

merveenoyan commented on examples/plot_tabular_regression.py:
```diff
@@ -0,0 +1,152 @@
+mae = mean_absolute_error(y_test, y_pred)
+mse = mean_squared_error(y_test, y_pred)
+r2 = r2_score(y_test, y_pred)
+model_card.add_metrics(**{"mean absolute error": mae, "mean squared error": mse, "r2 score": r2})
```
```suggestion
model_card.add_metrics(**{"Mean Absolute Error": mae, "Mean Squared Error": mse, "R-Squared Score": r2})
```

merveenoyan reviewed (approved):
Overall looks good, it's only nits, thank you!!!

lazarust commented:
@merveenoyan I think I've addressed all of your comments.

merveenoyan reviewed (approved):
Looks good to me! Thanks a lot!

merveenoyan commented on examples/plot_tabular_regression.py:
```diff
@@ -0,0 +1,149 @@
+model_card_authors = "skops_user, lazarust"
+citation_bibtex = "bibtex\n@inproceedings{...,year={2022}}"
+model_card.add(**{
+    "How to Get Started with the Model": get_started_code,
```
```suggestion
```
You can remove it as discussed 🙂

merveenoyan reviewed (commented):
Just noticed tests aren't passing :/

lazarust commented:
@merveenoyan Tests should be fixed now. I didn't realize that there was a check for formatting using Black.

merveenoyan commented:
@lazarust yesterday I saw the workflow failing for docs tests, as in, your script was failing, that's why.

merveenoyan commented on examples/plot_tabular_regression.py:
```diff
@@ -0,0 +1,154 @@
+Tabular Regression with scikit-learn
+-------------------------------------
+
+This example shows how you can create a Hugging Face Hub compatible repo for a
```
```suggestion
This example shows how you can use skops' capabilities to serialize a model, create a model card and a Hugging Face Hub repository for a
```

merveenoyan reviewed (approved):
LGTM!

merveenoyan commented:
@BenjaminBossan could you skim through it and see if it's ok to merge?

BenjaminBossan commented on docs/changes.rst:
```diff
@@ -83,4 +83,4 @@ Contributors
 :user:`Adrin Jalali <adrinjalali>`, :user:`Merve Noyan <merveenoyan>`,
 :user:`Benjamin Bossan <BenjaminBossan>`, :user:`Ayyuce Demirbas
-<ayyucedemirbas>`, :user:`Prajjwal Mishra <p-mishra1>`
+<ayyucedemirbas>`, :user:`Prajjwal Mishra <p-mishra1>`, :user:`Thomas Lazarus <lazarust>`
```
Should there not also be an entry (in a new section "v0.5") about what was added? Otherwise, it's hard to see the changes at a glance.

BenjaminBossan commented on examples/plot_tabular_regression.py:
```diff
@@ -0,0 +1,154 @@
+# Inference
+# =========
+# Let's see if the model works.
+prediction_data = X_test[:5]
+prediction = model.predict(prediction_data)
+print(prediction)
```
```suggestion
y_pred = model.predict(X_test[:5])
print(y_pred)
```

This is not a must, but I would prefer we use the canonical `y_pred` for the output of `model.predict` (further below, we also use `y_pred`). I know that in `plot_text_classification.py`, we also don't use `y_pred`, but I'd rather change it there as well, than keeping it here.

BenjaminBossan reviewed (changes requested):
Overall, this looks very good to me, thanks for the work. I have some minor comments, please take a look.

There is also a bigger issue: If I'm not mistaken, this example isn't linked anywhere in the docs. That means it would be really hard for readers to find it, they would have to manually change the URL to https://skops.readthedocs.io/en/stable/auto_examples/ to get there. Probably it would be best if we linked to the auto examples from some place, not necessarily to each auto example individually.

adrinjalali commented:
We can have a page where we manually list all examples in well defined categories for better discoverability.

BenjaminBossan commented:
> We can have a page where we manually list all examples in well defined categories for better discoverability.

Isn't that https://skops.readthedocs.io/en/stable/auto_examples/? We just need to link to it somewhere?

adrinjalali commented:
That file is auto generated and not really nice to read @BenjaminBossan , would be nicer for us to have a file which we maintain instead of linking to this autogenerated one.

BenjaminBossan commented:
> That file is auto generated and not really nice to read @BenjaminBossan , would be nicer for us to have a file which we maintain instead of linking to this autogenerated one.

Do you want that in this PR or can it be added in a separate one?

lazarust commented:
If we want to put it in this PR or a different one I can do it! It may be a couple days to a week till I can get to it, if that's alright.

lazarust commented:
@BenjaminBossan I've addressed your comments and cleaned up the commit history a bit.

For clarification on the new file for linking to the examples, are you just thinking it's a page (maybe at https://skops.readthedocs.io/en/stable/examples.html) that lists the example files and links to the example file on Github?

BenjaminBossan commented:
> I've addressed your comments and cleaned up the commit history a bit.

Thanks for that. Btw. no need to clean up the git history, because we squash before merging. Changing the git history can sometimes even make reviews harder.

> For clarification on the new file for linking to the examples, are you just thinking it's a page (maybe at https://skops.readthedocs.io/en/stable/examples.html) that lists the example files and links to the example file on Github?

IMO, let's start simple, as you suggested. Maybe add a sentence or two about what users can except and add a link to said page.

lazarust commented:
@BenjaminBossan I added the page to the documentation, but I don't super love the title and description of the examples. Let me know if you want me to change them!

BenjaminBossan commented on docs/examples.rst:
```diff
@@ -0,0 +1,9 @@
+.. _examples:
+
+Examples of using skops
```
I think the "of using skops" part is a bit redundant inside of the skops documentation ;)

adrinjalali replied:
```suggestion
Examples of interactions with the Hugging Face Hub
```

BenjaminBossan commented on docs/examples.rst:
```diff
@@ -0,0 +1,9 @@
+- Tabular Regression:
+    `Here <https://github.com/skops-dev/skops/blob/main/examples/plot_tabular_classification.py>`_ is an example of using skops to serialize a tabular regression model and create a model card and a Hugging Face Hub repository.
+- Text Classification:
+    `Here <https://github.com/skops-dev/skops/blob/main/examples/plot_text_classification.py>`_ is an example of using skops to serialize a text classification model and create a model card and a Hugging Face Hub repository.
```
I would expect all auto-examples to be listed here:

https://skops--254.org.readthedocs.build/en/254/auto_examples/

At the same time, the paragraph on the start page of the skops page becomes a bit redundant with this new page:

> The following examples are good starting points: ...

([link](https://skops.readthedocs.io/en/stable/index.html))

Should we remove it in favor of just linking to this page? @adrinjalali @merveenoyan WDYT?

adrinjalali replied:
I think that paragraph can stay since we can only link to a few examples where people can start, while also linking to this page with all examples listed.

lazarust replied:
@adrinjalali I'm a little confused, are we wanting to link from the start page to this examples page?

adrinjalali replied:
Yes we do @lazarust

adrinjalali commented on docs/examples.rst:
```diff
@@ -0,0 +1,9 @@
+Examples of using skops
+=======================
+
+- Tabular Regression:
+    `Here <https://github.com/skops-dev/skops/blob/main/examples/plot_tabular_classification.py>`_ is an example of using skops to serialize a tabular regression model and create a model card and a Hugging Face Hub repository.
+- Text Classification:
+    `Here <https://github.com/skops-dev/skops/blob/main/examples/plot_text_classification.py>`_ is an example of using skops to serialize a text classification model and create a model card and a Hugging Face Hub repository.
```
please keep the lines at max 79 chars

lazarust commented:
@adrinjalali @BenjaminBossan I've addressed your comments other than linking from the start page (if we want to do that).

BenjaminBossan commented on docs/examples.rst:
```diff
@@ -0,0 +1,21 @@
+    regression model and create a model card and a Hugging Face Hub repository.
+- Text Classification:
+    `Here <https://github.com/skops-dev/skops/blob/main/examples/plot_text_cl
+    assification.py>`_ is an example of using skops to serialize a text classi
```
You accidentally split the word "classification" into 2.

BenjaminBossan commented on docs/examples.rst:
```diff
@@ -0,0 +1,21 @@
+    _classification.py>`_ is an example of using skops to serialize a tabular 
+    regression model and create a model card and a Hugging Face Hub repository.
+- Text Classification:
+    `Here <https://github.com/skops-dev/skops/blob/main/examples/plot_text_cl
```
All the links here are to the .py files on GH. However, it is better to point them to the rendered documentation. E.g. for text classification, that would be: https://skops.readthedocs.io/en/stable/auto_examples/plot_text_classification.html

lazarust commented:
@BenjaminBossan I've made the updates you requested, but it looks like there's an error happening in `plot_hf_hub.py` when it uploads the model to the Hub. Would you be able to help me figure out why this is happening?

BenjaminBossan commented:
@lazarust I have no idea, it seems unlikely that any change you made could cause that. I tried to trigger a rebuild of the docs in case the problem was just temporary, but didn't see an option to do that. Could you please push an empty commit to trigger it?

adrinjalali commented:
I triggered a rebuild on the readthedocs side and it passes

adrinjalali commented:
@lazarust please avoid force pushing to your branch. We squash and merge at the end, so it doesn't matter how many commits with what commit messages you have in between.

lazarust commented:
@adrinjalali Should I update the branch via rebase or merge commit?

lazarust commented:
@BenjaminBossan This is ready for you to review again

BenjaminBossan commented on docs/examples.rst:
```diff
@@ -0,0 +1,17 @@
+    :ref:`sphx_glr_auto_examples_plot_hf_hub.py` is an example of using skops 
+    to put a model card on the Hugging Face Hub.
+- Tabular Regression:
+    :ref:`sphx_glr_auto_examples_plot_tabular_regresssion.py` is an example of using skops to serialize a tabular 
```
```suggestion
    :ref:`sphx_glr_auto_examples_plot_tabular_regression.py` is an example of using skops to serialize a tabular 
```

BenjaminBossan reviewed (commented):
Just this typo, then we should be good :)

lazarust commented:
@BenjaminBossan Fixed!

BenjaminBossan reviewed (approved):
Great, thanks a lot for working on this!
